File: backend/global-service/src/main/java/com/nook/global_service/service/AuthService.java
package com.nook.global_service.service;

import java.util.Objects;

import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import com.nook.global_service.common.ResponseDto;
import com.nook.global_service.common.dto.AuthUserDto;
import com.nook.global_service.common.dto.CreateUserDto;
import com.nook.global_service.common.exception.CommonException;
import com.nook.global_service.entity.UserEntity;
import com.nook.global_service.repository.UserRepository;

@Service
public class AuthService {
    private final UserRepository userRepository;

    AuthService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public ResponseDto createUser(CreateUserDto request) {
        UserEntity userEntity = UserEntity.builder().email(request.email()).password(request.password()).username(request.username()).build();
        
        if(Objects.nonNull(userRepository.existsWithEmail(request.email()))) {
            throw new CommonException(HttpStatus.BAD_REQUEST, "Email already exists: " + request.email());
        }
        
        userRepository.save(userEntity);

        return new ResponseDto(HttpStatus.CREATED, "User created successfully", null);
    }

    public ResponseDto authUser(AuthUserDto request) {
        UserEntity userEntity = userRepository.authenticateUser(request.email(), request.password());

        if(Objects.isNull(userEntity)) {
            throw new CommonException(HttpStatus.UNAUTHORIZED, "Invalid email & password combination.");
        }

        return new ResponseDto(HttpStatus.OK, "User logged in: " + request.email(), null);
    }
}

File: src/modules/auth/auth.service.ts
import {
  BadRequestException,
  HttpStatus,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { CreateDto } from './dto/create.dto';
import { InjectRepository } from '@nestjs/typeorm';
import { UserEntity } from 'src/entities/user.entity';
import { Repository } from 'typeorm';
import { LoginDto } from './dto/login.dto';
import { ResponseInterface } from 'src/common/interfaces/response.interface';
import { SessionInterface } from 'src/common/interfaces/session.interface';
import * as bcrypt from "bcrypt";
import { AlbumsEntity } from 'src/entities/albums.entity';
import { MediaEntity } from 'src/entities/media.entity';
import * as uuid from "uuid"

@Injectable()
export class AuthService {

  saltRound = 10;

  constructor(
    @InjectRepository(UserEntity)
    private userRepository: Repository<UserEntity>,
    @InjectRepository(AlbumsEntity)
    private albumRepository: Repository<AlbumsEntity>
  ) {}

  async createUser(createDto: CreateDto): Promise<ResponseInterface> {
    const { email, password, username } = createDto;

    if(await this.userRepository.existsBy({email, is_active: false}))
      throw new UnauthorizedException(HttpStatus.UNAUTHORIZED, "Your account is currently disabled or banned. Please contact support if you believe this is a mistake.")

    if (await this.userRepository.existsBy({ email }))
      throw new BadRequestException(HttpStatus.BAD_REQUEST,'A user with this email already exists. Please use a different email address or log in if you already have an account.');

    
    const hash = await bcrypt.hash(password, this.saltRound);

    let user = new UserEntity();
    user.id = uuid.v4();
    user.email = email;
    user.username = username;
    user.password = hash;

    let album: AlbumsEntity = {
      id: uuid.v4(),
      albumName: "Favorite",
      userId: user.id,
      content: []
    };

    user.favoriteAlbumID = album.id;

    await this.userRepository.save(user);
    await this.albumRepository.save(album);
    
    return {
      statusCode: HttpStatus.CREATED,
      message: 'Your account has been created successfully!'
    };
  }

  async loginUser(loginDto: LoginDto, session: SessionInterface): Promise<ResponseInterface> {
    const { email, password } = loginDto;

    const user: UserEntity | null = await this.userRepository.findOneBy({
      email, is_active: true
    });

    if (!user) {
      throw new BadRequestException(
        HttpStatus.BAD_REQUEST,
        'No active account found with this email address.',
      );
    }

    if(await bcrypt.compare(password, user.password)===false) {
      throw new UnauthorizedException(HttpStatus.UNAUTHORIZED, 'Incorrect email or password. Please try again.');
    }

    session.userId = user.id;

    return {
      statusCode: HttpStatus.OK,
      message: `Login successful! Welcome back, ${user.username}.`,
    };
  }
}
